scene/main/canvas_item.cpp

```diff
@@ -92,11 +96,13 @@ void CanvasItem::set_visible(bool p_visible) {
 	visible = p_visible;
 	RenderingServer::get_singleton()->canvas_item_set_visible(canvas_item, p_visible);
 
-	if (!is_inside_tree()) {
+	if (!parent_visible_in_tree) {
+		notification(NOTIFICATION_VISIBILITY_CHANGED);
 		return;
 	}
 
 	_propagate_visibility_changed(p_visible, !p_visible);
+	parent_visible_in_tree = true; // Keep it true after propagation.
```

Comment:
I still think it'll be more accurate and you won't need to do manual reverting like this if we change the second parameter of `_propagate_visibility_changed()` to `p_is_source` :P

Reply:
Yeah, on another look, it makes sense to always pass `true` here.